src/riak_kv_qry_worker.erl

```diff
@@ -566,4 +590,46 @@ prepare_final_results_test() ->
                 result = [{1, Rows}]})
     ).
 
+estimate_query_size_limit_applies_to_aggregating_queries_test() ->
+    BigData = [<<"BIGFATDATA">>],
+    check_states(
+      #state{qry = ?SQL_SELECT{'SELECT' = #riak_sel_clause_v1{calc_type = group_by}},
+             result            = BigData,
+             max_query_data    = erlang:external_size(BigData) - 1,
+             sub_qrys          = lists:seq(1, 100)}).
+
+estimate_query_size_limit_applies_to_regular_queries_test() ->
+    check_states(
+      #state{qry = ?SQL_SELECT{'SELECT' = #riak_sel_clause_v1{calc_type = rows}},
+             total_query_data  = 10,
+             max_query_data    = 30,
+             sub_qrys          = lists:seq(1, 100)}).
+
+check_states(State) ->
+    lists:foreach(
+      fun({StateN, Outcome}) -> ok = check_states2(StateN, Outcome) end,
+      [{State#state{n_subqueries_done = 1}, passing},
+       {State#state{n_subqueries_done = 2}, cancelled},
+       {State#state{n_subqueries_done = 3}, cancelled}]).
+
+check_states2(State, passing) ->
+    ?assertEqual(estimate_query_size(State), State);
+check_states2(State, cancelled) ->
+    ErrorReceiverPid =
+        erlang:spawn(
+          fun() -> receive {error, select_result_too_big} -> ok end end),
+    ?assertEqual(estimate_query_size(State#state{receiver_pid = ErrorReceiverPid}), new_state()),
+    ok = check_error_receiver_self_destructs(ErrorReceiverPid, 50).
+
+check_error_receiver_self_destructs(_Pid, 0) ->
+    didnt_receive_error_message;
+check_error_receiver_self_destructs(Pid, N) ->
+    case is_process_alive(Pid) of
+        false ->
+            ok;
+        true ->
+            timer:sleep(100),
```

Comment:
This `time:sleep/1` is making the test very slow.

Preferably the side effect could be moved out of the function being tested.

Alternatively, the test process should receive a message instead of waiting on this process to die.